java/flight/src/main/java/org/apache/arrow/flight/FlightService.java

```diff
@@ -191,17 +191,17 @@ public void completed() {
             StreamPipe.wrap(responseObserver, PutResult::toProtocol)).run();
         responseObserver.onCompleted();
       } catch (Exception ex) {
-        logger.error("Failed to process custom put.", ex);
         responseObserver.onError(ex);
-        // The client may have terminated, so the exception here is effectively swallowed.
         // Log the error as well so -something- makes it to the developer.
         logger.error("Exception handling DoPut", ex);
```

Comment:
generally we should be throwing or logging, not both, is there a strong reason to do both here?

Reply:
@emkornfield good question.

I think there are two reasons:

1. IMO, printing a log is not an appropriate way of handling an exception. The proper way should depend on the concrete exception, and we may decide to ignore it silently, print a log, or terminate the server, etc.

2. For some scenario, the logger may not be properly set, which makes debugging extremely difficult. We have such a experience when resolving ARROW-5868. It seems this case is also prevalent in customer settings.

About 1), it should be noted that, the exception thrown during a call to submit is silently swallowed due to the mechanism of Java thread pool. However, I think this is the first step towards the right path.

Reply:
> About 1), it should be noted that, the exception thrown during a call to submit is silently swallowed due to the mechanism of Java thread pool. However, I think this is the first step towards the right path.

Can you outline subsequent steps.  If the exception isn't propagated outside the pool, it seems like this isn't super useful?

Reply:
> > About 1), it should be noted that, the exception thrown during a call to submit is silently swallowed due to the mechanism of Java thread pool. However, I think this is the first step towards the right path.
> 
> Can you outline subsequent steps. If the exception isn't propagated outside the pool, it seems like this isn't super useful?

We can extend the ThreadExecutor to catch the exceptions. However, a more difficult problem arises: how to deal with each particular exception? When do we need to terminate the service/server? When do we need to neglect them?

Reply:
So this is why I originally asked above; any exception will just bubble up to gRPC's executor which will just also log it and discard it. There isn't much the developer can do at that point.

Reply:
@lihalite thanks for your comments.

The executor is provided by FlightService, so in theory, it is possible to extend the ThreadPoolExecutor class and override the afterExecute to process the exception.

Reply:
I think I would prefer not to throw an exception until we know how it is going to be used.  Another alternative is for the service to take a callback to record these types of things.  @lihalite I think you should consider the alternatives as you implement middleware for monitoring/etc.  It would also likely be nice to let instantiators of the service pass in the thread-pool they want to use.

Reply:
Yes, it's probably best to send errors to any middleware and let them log as they please. (They won't be able to communicate with the client, but that's fine as the client has likely already disconnected.) However, we should make sure something makes it to the developer by default, I originally added the log line because the exception was getting silently swallowed, making debugging frustrating.

Reply:
Hmm, that executor does need to be a parameter. The only caveat is that it needs to be a distinct executor from the gRPC executor, or you can deadlock the server.

Reply:
Right so to close this off, leave the log and remove throwing the exception and figure out how to propogate the exception as part of middleware.